### DISCUSSION TITLE
Schema validation on service startup?

### DISCUSSION THREAD

**Discussion by damar-block** [skeema/skeema#233] [Q&A] [answered]
Today our services use [Flyway](https://www.baeldung.com/database-migrations-with-flyway) to manage DB migration. Flyway manages a migrations table which basically stores the version of applied migration.

When service starts, we compare the migrations available in the deployed build to the migrations applied on DB. If there is any migration in the build that isn't applied on DB - we fail the service startup. The idea is to prevent running a new build that contains an un-applied migration (and thus might rely on it).

What would be the equivalent validation to do using Skeema on service startup? Skeema's approach is different - there is no longer order and versioning between schema changes. Is there any existing experience we can learn from?

Thanks!

**Comment by evanelias** [comment, marked as answer]
The simplest approach would be to run [`skeema diff`](https://www.skeema.io/docs/commands/diff/) and see if the exit code is 0 (no differences) or 1+ (differences found, meaning the schema repo has at least one change that hasn't been pushed to the db server yet). For more info, please see the doc page on [automation](https://www.skeema.io/docs/automation/), and also some of the [recipes](https://www.skeema.io/docs/recipes/) might be helpful.

re: "no longer order and versioning between schema changes", yes and no... Skeema is 100% intended to be used with a version control system, such as Git. The core idea is that schema changes should be managed and versioned exactly like any other codebase: each company can decide how they want to manage branches, tags, pull requests, CI/CD etc, and Skeema helps enable that for schema changes, but isn't opinionated itself in how you do versioning.

In contrast, traditional imperative migration tools (such as Flyway) enforce a versioning system that must be stored in a db table and often also applied as prefixes to file or dir names, but this is extraneous to the completely separate versioning in Git: there's no guarantee that their incremental migration versioning scheme lines up with Git commit ordering. This means that with a traditional migration tool, there's multiple sources of truth which can diverge. That's a bad thing :)

With Skeema, if you need incremental schema versioning and more complex gating logic around application deploys -- e.g. the ability to say "this application change depends on a specific schema change/PR" -- this can be implemented using Git tags, among other ways. See [the bottom half of this comment](https://github.com/cashapp/spirit/issues/292#issuecomment-2148802204) for three possible Git-native approaches.

Also wanted to note that while Skeema itself isn't tied to Git specifically, a separate Git-oriented product might be introduced in the future, which would provide more of this functionality out-of-the-box.

**Comment by evanelias** [comment]
Closing this out, but please let me know if you have any follow-up questions, happy to reopen if so.

(Side note: I'm also considering enabling GitHub's separate "discussions" feature for this repo sometime soon, as that might be a better forum for casual questions -- no need to "close" questions there...)
